runtime/runtime-async/src/commonMain/kotlin/app/cash/sqldelight/async/AsyncTransacter.kt

```diff
@@ -0,0 +1,267 @@
+}
+
+private class AsyncTransactionWrapper<R>(
+  val transaction: AsyncTransacter.Transaction
+) : AsyncTransactionWithoutReturn, AsyncTransactionWithReturn<R> {
+  override fun rollback(): Nothing {
+    transaction.checkThreadConfinement()
+    throw RollbackException()
+  }
+
+  override fun rollback(returnValue: R): Nothing {
+    transaction.checkThreadConfinement()
+    throw RollbackException()
+  }
+
+  /**
+   * Queues [function] to be run after this transaction successfully commits.
+   */
+  override fun afterCommit(function: () -> Unit) {
+    transaction.afterCommit(function)
+  }
+
+  /**
+   * Queues [function] to be run after this transaction rolls back.
+   */
+  override fun afterRollback(function: () -> Unit) {
+    transaction.afterRollback(function)
+  }
+
+  override suspend fun transaction(body: suspend AsyncTransactionWithoutReturn.() -> Unit) {
+    transaction.transacter!!.transaction(false, body)
+  }
+
+  override suspend fun <R> transactionWithResult(body: suspend AsyncTransactionWithReturn<R>.() -> R): R {
+    return transaction.transacter!!.transactionWithResult(false, body)
+  }
+}
+
+internal class RollbackException(val value: Any? = null) : Throwable()
+
+/**
+ * A transaction-aware [AsyncSqlDriver] wrapper which can begin a [Transaction] on the current connection.
+ */
+abstract class AsyncTransacterImpl(private val driver: AsyncSqlDriver) : AsyncTransacter {
+  /**
+   * For internal use, notifies the listeners provided by [listenerProvider] that their underlying result set has
+   * changed.
+   */
+  protected fun notifyQueries(identifier: Int, tableProvider: ((String) -> Unit) -> Unit) {
+    val transaction = driver.currentTransaction()
+    if (transaction != null) {
+      if (transaction.registeredQueries.add(identifier)) {
+        tableProvider { transaction.pendingTables.add(it) }
+      }
+    } else {
+      val tableKeys = mutableSetOf<String>()
+      tableProvider { tableKeys.add(it) }
+      driver.notifyListeners(tableKeys.toTypedArray())
+    }
+  }
+
+  /**
+   * For internal use, creates a string in the format (?, ?, ?) where there are [count] offset.
+   */
+  protected fun createArguments(count: Int): String {
+    if (count == 0) return "()"
+
+    return buildString(count + 2) {
+      append("(?")
+      repeat(count - 1) {
+        append(",?")
+      }
+      append(')')
+    }
+  }
+
+  override suspend fun transaction(
+    noEnclosing: Boolean,
+    body: suspend AsyncTransactionWithoutReturn.() -> Unit
+  ) {
+    return transactionWithWrapper(noEnclosing, body)
+  }
+
+  override suspend fun <R> transactionWithResult(noEnclosing: Boolean, bodyWithReturn: suspend AsyncTransactionWithReturn<R>.() -> R): R {
+    return transactionWithWrapper(noEnclosing, bodyWithReturn)
+  }
+
+  private suspend fun <R> transactionWithWrapper(noEnclosing: Boolean, wrapperBody: suspend AsyncTransactionWrapper<R>.() -> R): R {
+    val transaction = driver.newTransaction()
```

Comment:
probably we should figure out how to share this logic between the sync/async transacter types, but we can do that in a follow up